src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs

```diff
@@ -414,59 +426,50 @@ protected virtual void ValidateIdToken(OpenIdConnectProtocolValidationContext va
+        /// <returns>A <see cref="HashAlgorithm"/>.</returns>
         public virtual HashAlgorithm GetHashAlgorithm(string algorithm)
         {
             if (algorithm == null)
-            {
                 algorithm = SecurityAlgorithms.RsaSha256;
-            }
 
             try
             {
-                switch (algorithm)
-                {
-                    case SecurityAlgorithms.Sha256:
-                    case SecurityAlgorithms.EcdsaSha256:
-                    case SecurityAlgorithms.HmacSha256:
-                    case SecurityAlgorithms.RsaSha256:
-                    case SecurityAlgorithms.RsaSsaPssSha256:
-                    case SecurityAlgorithms.RsaSha256Signature:
-                    case SecurityAlgorithms.EcdsaSha256Signature:
-                        return SHA256.Create();
-
-                    case SecurityAlgorithms.Sha384:
-                    case SecurityAlgorithms.EcdsaSha384:
-                    case SecurityAlgorithms.HmacSha384:
-                    case SecurityAlgorithms.RsaSha384:
-                    case SecurityAlgorithms.RsaSsaPssSha384:
-                    case SecurityAlgorithms.RsaSha384Signature:
-                    case SecurityAlgorithms.EcdsaSha384Signature:
-                        return SHA384.Create();
-
-                    case SecurityAlgorithms.Sha512:
-                    case SecurityAlgorithms.RsaSha512:
-                    case SecurityAlgorithms.EcdsaSha512:
-                    case SecurityAlgorithms.HmacSha512:
-                    case SecurityAlgorithms.RsaSsaPssSha512:
-                    case SecurityAlgorithms.RsaSha512Signature:
-                    case SecurityAlgorithms.EcdsaSha512Signature:
-                        return SHA512.Create();
-                }
+                string hashAlgorithm;
+                if (!HashAlgorithmMap.TryGetValue(algorithm, out hashAlgorithm))
+                    hashAlgorithm = algorithm;
 
+                return CryptoProviderFactory.CreateHashAlgorithm(hashAlgorithm);
             }
             catch (Exception ex)
             {
-                throw LogHelper.LogException<OpenIdConnectProtocolException>(ex, LogMessages.IDX10301, algorithm);
+                throw LogHelper.LogException<OpenIdConnectProtocolException>(ex, LogMessages.IDX10301, algorithm, typeof(HashAlgorithm));
             }
 
             throw LogHelper.LogException<OpenIdConnectProtocolException>(LogMessages.IDX10302, algorithm);
         }
 
         /// <summary>
+        /// Gets or sets the <see cref="CryptoProviderFactory"/> that will be used for crypto operations.
+        /// </summary>
+        public CryptoProviderFactory CryptoProviderFactory
+        {
+            get
+            {
+                return _cryptoProviderFactory;
+            }
+            set
+            {
+                if (value == null)
+                    throw LogHelper.LogArgumentNullException(nameof(value));
```

Comment:
Nothing wrong with it, but since the name of the implicit "value" param will never change names, seems more performant to avoid the lookup and just provide the string "value". nameof provides consistency across parameter renames that we don't need on getters/setters.